### DISCUSSION TITLE
Pore network model from CT scan

### DISCUSSION THREAD

**Discussion by nuritg** [PMEAL/porespy#605] [Q&A] [answered]
Dear Prof. Gostick, 

I am a Ph.D student from the Hebrew University of Jerusalem, Israel, interested in porous media, and I'm trying to generate a 2D pore network from my own CT scans, in order to get the fitting water retention curve from the image (that's for now, later on, I will try also the 3D network and some other good things you offer :)), however, after reducing noise from the image I'm getting this result, which is quite good I think, but there are a few large pores that are left empty. what am I doing wrong? Or do you have an idea how to make it better (also for the noise reduction)? I'm a bit new to Python...

- One other thing, I used the pore volume as pore area since I found that the pore area does not make sense here, maybe it fits only to the 3D?

Lots of thanks.
Nurit

Attached is the code:
![0606_image](https://user-images.githubusercontent.com/36815224/146680399-8fb883d7-0ac2-4795-98c7-3426f93a1c08.jpg)
![pore_networkCT606](https://user-images.githubusercontent.com/36815224/146680837-77f830c3-f32e-4902-8a4b-4fe53151cd20.JPG)
 
```python
import skimage.viewer
import matplotlib.pyplot as plt

from skimage import data
from skimage import color, morphology

import numpy as np
import porespy as ps
import openpnm as op


image = skimage.io.imread(fname=r'C:\Users\nuritg\Desktop\0606_image.jpg', as_gray=True)

#show image:
plt.figure(1)
plt.imshow(image, cmap='gray')
plt.show()

#crop image:
print(image.shape)
image = image[500:1500, 500:1500]
#image = image[700:3000, 700:3200]
plt.show()
print(image.shape)
plt.imshow(image, cmap='gray')


Otsus_threshold = skimage.filters.threshold_otsu(image)


plt.figure(1)
mask_image = image<Otsus_threshold
plt.imshow(mask_image, cmap='gray')
plt.show()


plt.figure(3)
selem =  morphology.disk(1)
image_2 = morphology.opening(mask_image, selem)
selem =  morphology.disk(4)
image_2 = morphology.closing(image_2, selem)

plt.imshow(image_2, cmap='gray')
plt.show()

im = image_2
plt.figure(figsize=(6, 6))
plt.imshow(im)
plt.axis('off')
plt.show()
snow_output = ps.networks.snow(
    im,
    voxel_size=1,
    boundary_faces=['left','right','front','back'],
    marching_cubes_area=False
)

pn = op.network.GenericNetwork()

pn.update(snow_output)
pn['pore.true_diameter']= (pn['pore.volume']/np.pi)**0.5
prj = pn.project

prj.export_data(filename='extraction_im606', filetype='vtk')
ps.io.to_vtk(np.array(im, dtype=int)[:, :, np.newaxis], 'im606')
```

**Comment by jgostick** [comment, marked as answer]
When you say 'left empty' are you referring to the pores with no spheres in it?  If so, there is a setting in paraview where it only shows a limited number of glyphs, which is on by default.  You have to set it to something like 'all points'.  Its a drop down in the properties panel.  Other than that, the image looks great!

Regarding the 2D vs 3D thing, you'll want to do it in 3D for sure, even if you only do a fraction of the total slices.  The first reason is that connectivity is much different in 2D vs 3D.  In your picture above there are many pores that are like dangling dead-ends.  Adding the 3rd dimension will cause these to connect with each other.  Secondly, as you point out, meaning of volume and other geometric properties becomes lost in 2D.

**Reply by nuritg** [reply]
Thanks for the quick response!
It worked :) and the picture looks great.
